### PR TITLE
Value parsing

### PR DESCRIPTION
**Pull request by robacourt**
Closes https://github.com/electric-sql/electric-next/issues/164.

### CODE REVIEW

**Comment by kevin-dp** [comment]
This PR parses values of type `int2`, `int4`, `int8`, `bool`, `float8`, `json`, and `jsonb` on the client.

File: packages/typescript-client/src/client.ts [hunk 1]
```diff
@@ -207,53 +212,60 @@ export class ShapeStream {
...
-          })
-          .then((batch: Message[]) => {
-            // Update isUpToDate & lastOffset
-            if (batch.length > 0) {
-              const lastMessages = batch.slice(-2)
+        const messages = status === 204 ? `[]` : await response.text()
+        const batch = JSON.parse(messages, (key, value) => {
+          if (key === `value`) {
+            // Parse the row values
+            const row = value as Record<string, Value>
```
Comment (icehaunter) [inline packages/typescript-client/src/client.ts:241]: this doesn't work correctly when there is a column named `value` - it's a good idea to actually add a test with a column named that. We prob need to `typeof value == 'object'` here

File: packages/typescript-client/src/client.ts [hunk 2]
```diff
@@ -207,53 +212,60 @@ export class ShapeStream {
...
+          }
+          return value
+        }) as Message[]
 
-              lastMessages.forEach((message) => {
-                if (message.headers?.[`control`] === `up-to-date`) {
-                  const wasUpToDate = this.isUpToDate
+        // Update isUpToDate & lastOffset
+        if (batch.length > 0) {
+          const lastMessages = batch.slice(-2)
```
Comment (icehaunter) [inline packages/typescript-client/src/client.ts:251]: since you're refactoring, this no longer updates `lastOffset` at all, given we're doing it from headers now. So we can just check the last element for up-to-date instead of slicing last 2 and iterating over them,=

File: packages/typescript-client/src/client.ts [hunk 3]
```diff
@@ -207,53 +212,60 @@ export class ShapeStream {
       }
 
       try {
-        await this.fetchClient(url.toString(), { signal })
-          .then(async (response) => {
-            if (!response.ok) {
-              throw await FetchError.fromResponse(response, url.toString())
-            }
+        const response = await this.fetchClient(url.toString(), { signal })
```
Comment (icehaunter) [inline packages/typescript-client/src/client.ts:215]: I love the `await` refactoring here! One improvement that would be nice to get in to make iteration over this easier maybe would be to make error handling a bit more careful? I mean to only wrap the `fetch` call in a try-catch (and maybe inside a separate function?) and to only handle fetch errors there, so that we're not blindly retrying on failure to parse a type or smth like that

File: packages/typescript-client/src/client.ts [hunk 4]
```diff
@@ -388,6 +401,37 @@ export class ShapeStream {
...
+    const columnInfo = this.schema?.[key]
+    // use guards for the "type" property
+    if (!columnInfo) {
+      // We don't have information about the value
+      // so we just return it
+      return value
+    }
+
+    // Pick the right parser for the type
+    const parser =
```
Comment (icehaunter) [inline packages/typescript-client/src/client.ts:416]: nitpick: correct type here is

```suggestion
    const parser: ParseFunction | undefined =
```

File: packages/typescript-client/src/client.ts [hunk 5]
```diff
@@ -388,6 +401,37 @@ export class ShapeStream {
...
+      this.parser?.[columnInfo.type] ?? defaultParser[columnInfo.type]
```
Comment (icehaunter) [inline packages/typescript-client/src/client.ts:417]: Can we make this cheaper by merging the provided parser with default parser in the constructor?

File: packages/typescript-client/test/parser.test.ts [hunk 6]
```diff
@@ -0,0 +1,98 @@
...
+      1, 2, 3, 4, 5,
+    ])
+    expect(pgArrayParser(`{1,2,3,4,5}`, defaultParser.int8)).toEqual([
+      BigInt(1),
+      BigInt(2),
+      BigInt(3),
+      BigInt(4),
+      BigInt(5),
+    ])
+    expect(pgArrayParser(`{"foo","bar"}`, (v) => v)).toEqual([`foo`, `bar`])
```
Comment (icehaunter) [inline packages/typescript-client/test/parser.test.ts:72]: How about an addiitonal test over real edge cases?
```suggestion
    expect(pgArrayParser(`{"foo","bar"}`, (v) => v)).toEqual([`foo`, `bar`])
    expect(pgArrayParser(`{foo,"}"}`, (v) => v)).toEqual([`foo`, `}`])
```

**Review by icehaunter — CHANGES_REQUESTED** [review]
Great work! I found a bug that definitely needs addressing before merge, and some nitpicks

File: packages/typescript-client/src/client.ts (hunk 1 above)
Comment (kevin-dp) [reply packages/typescript-client/src/client.ts:241]: Indeed, great catch!

File: packages/typescript-client/src/client.ts (hunk 5 above)
Comment (kevin-dp) [reply packages/typescript-client/src/client.ts:418]: Good idea :-)

File: packages/typescript-client/src/parser.ts [hunk 7]
```diff
@@ -0,0 +1,74 @@
...
+  additionalInfo?: Omit<ColumnInfo, `type` | `dims`>
+) => Value
+export type Parser = { [key: string]: ParseFunction }
+
+const parseNumber = (value: string) => Number(value)
+const parseBool = (value: string) => value === `true` || value === `t`
+const parseBigInt = (value: string) => BigInt(value)
+const parseJson = (value: string) => JSON.parse(value)
+
+export const defaultParser: Parser = {
```
Comment (KyleAMathews) [inline packages/typescript-client/src/parser.ts:14]: no dates? It seems like most (all?) postgres dates could be converted to a Date object.

File: packages/typescript-client/test/integration.test.ts [hunk 8]
```diff
@@ -163,6 +163,202 @@ describe(`HTTP Sync`, () => {
...
+        value: { a: 5, b: [{ c: `foo` }] },
+      },
+    ])
+
+    // Now update the data
+    await dbClient.query(
+      `
+      UPDATE ${table}
+      SET
+        txt = 'changed',
```
Comment (KyleAMathews) [inline packages/typescript-client/test/integration.test.ts:291]: looking at this custom serializer/parser — tests for `Infinity` & `NaN` would be good as they're supported in both postgres and js https://github.com/Rich-Harris/devalue

**Review by KyleAMathews — COMMENTED** [review]
Looks great!

**Comment by KyleAMathews** [comment]
Running parsing a big value through Chrome's performance tracer would be good too as a sanity check we're doing things as cheaply as possible.

File: packages/typescript-client/test/parser.test.ts (hunk 6 above)
Comment (kevin-dp) [reply packages/typescript-client/test/parser.test.ts:80]: I added some corner cases, you can have a look to see if it covers them all or if you want some additional ones

File: packages/typescript-client/src/parser.ts (hunk 7 above)
Comment (kevin-dp) [reply packages/typescript-client/src/parser.ts:14]: I think we want to stay away from Dates as turning PG dates into JS dates can be confusing.
If you have a timestamp with time zone it's easy, but if you have for instance just a postgres `Date` then what do you do? Do you turn it into a JS date and set the time to midnight? Midnight local time or midnight UTC? 

Similarly, if you have a PG `time` without a date, what do you do? You add today's date? You add the first epoch day (1st jan. 1970)?

What do you do with Postgres timestamps without time zone? Do you parse them as that time in the client's current time zone?

I think it's best we are not opinionated and let clients decide how they want to handle it.
At most, i would add timestamps with time zones as those are pretty clear how to handle.

File: packages/typescript-client/test/integration.test.ts (hunk 8 above)
Comment (kevin-dp) [reply packages/typescript-client/test/integration.test.ts:264]: Good idea, will add!

File: packages/typescript-client/src/parser.ts (hunk 7 above)
Comment (KyleAMathews) [reply packages/typescript-client/src/parser.ts:14]: cool yeah makes sense. And people can easily add their own custom parsing now so if it's straightforward to get Date objects, then they can do it.

**Comment by kevin-dp** [comment]
Plot of a quick benchmark comparing our value parser vs built-in `JSON.parse`

Screenshot 2024-07-30 at 12 50 22

**Comment by KyleAMathews** [comment]
I don't love extra CPU time but... the time is still dominated by server/network time so it's not a big deal and we can still move parsing to the server for extra client optimization in the future, if needed.

**Review by KyleAMathews — APPROVED** [review]
:shipit:

**Review by icehaunter — APPROVED** [review]
